### ISSUE TITLE
Bounty - Integrate/Bundle Level Designer Toolkit (LDtk) into GDevelop as a Tilemap editor

### ISSUE THREAD

**Issue by Silver-Streak** [4ian/GDevelop#2991]
## Description

This post is to allow for a bounty to be opened/reopened on the effort discussed here: 
https://github.com/4ian/GDevelop/issues/2002
https://github.com/4ian/GDevelop/discussions/2352
https://github.com/deepnight/ldtk/issues/126

LDtk is a modern Tilemap editor and world builder made by deepknight, one of the developers behind Dead Cells.

It is a lightweight HTML5 based Tilemap editor, with numerous advanced features such as Automapping, advanced entity support for logic/objects/etc, and embedding the tilesheet into the final tilemap file.

![94418415-4aa61100-0179-11eb-9aee-0451fe67b296](https://user-images.githubusercontent.com/24237149/131237236-76069458-1753-495d-b058-ccc4bf9b9775.gif)

The main site is here: https://ldtk.io
The github repo is here: https://github.com/deepnight/ldtk

## Solution suggested

Based off the above discussion, the goal would be for this integration to include the following: 
- Wrapping LDtk as a react component for GDevelop
- Adding the react component as a new tab in GDevelop 5.
- Allow for two-way communication between GD5 and LDtk, or at least for edits to existing tilemaps in LDtk to update tilemap objects in GD5 (may already be covered by the below PR)

## Alternatives considered

- LDtk tilemap support in the tilemap object was an original alternative suggested for the first issue on this. This is currently being implemented in https://github.com/4ian/GDevelop/pull/2828 , and helps with a lot of the logical items.
- Building out a custom GDevelop tilemap editor is another option, but that seems to replicate work done elsewhere, and means that a lot of the development would have to be helmed by the GDevelop contributors long term, whereas the majority of functional and logical needs behind tilemaps are already covered by LDtk.

I am pinging @4ian and @blurymind to review the above and explicitly the bullets in the solution suggested. Once everyone agrees, I'll work to get the new bounty opened and the old funds moved. (This will not impact anything with the current bounty on #2828)

**Comment by 4ian** [comment]
I've added "If too complex, consider running LDtk as a separate dialog (like Piskel)" to the potential solution. This is a quite open ended issue and if someone wants to tackle this, I'll be interested in a solution that is both simple to maintain (easy to upgrade LDtk without changes, clear interface) and has a good UX (we're not supporting everything that LDtk allows in our Tilemap object -> should this be disabled in LDtk interface?).

**Comment by Silver-Streak** [comment]
I've started the claim process to reclaim the funds. Unfortunately since I'm one of the bounty funders, it will not allow me to approve it. It'll have to wait for 9/18 before I can reclaim them, and I'll move it over to resolv at that point.

**Comment by blurymind** [comment]
I can do it easily as an external editor, but will first need to get the
parsing merged in the other pr.

I think the bundling part of the job is actually pretty straightforward. It
shouldn't take me longer than a day.

**Comment by Silver-Streak** [comment]
@blurymind What would be the likelihood of going for the full integration rather than bundling? 

Unless I'm misunderstanding, including it as an external editor would make it harder to utilize entities/reflect changes as they happen, right?

**Comment by blurymind** [comment]
Well it would be harder to add it as a component, but will also mean the
editor will work in gdevelops web version.

As far as bundling it as an external one - it will be similar to piskel if
you want me to do it that way. We have to remember that ldtk itself is not
designed to work as a web app and its natural wrapper is electron - so
going the external editor route would probably be a safer bet.
Basically you would only be able to have one instanceof it open and it will
be modal - so while you use it, you cant use other parts of gdevelop. When
done, you apply your changes and they get written to files

**Comment by Silver-Streak** [comment]
Do you think that bundling would limit/restrict our ability to utilize entities/enums from LDtk vs integrating?

**Comment by blurymind** [comment]
Not necessarily.  I see that more as another problem. Entities and enums
are still a part of the tilemap object, which you are only editing in the
bundled ldtk editor. How that data is utilised by gdevelop can depend on
what you have done with it in the event sheet. Remember that gdevelop can
have methods of reading from it to create game objects, etc when the game
runs.

**Comment by Silver-Streak** [comment]
I would _really_ like to see Entity and Enum data be available as part of this bounty, or at least Entity. Does that seem viable? 

If so then I think that path forward seems great.

**Comment by blurymind** [comment]
I can try to add setters/ getters to do it in the current ldtk parser pr

**Comment by Silver-Streak** [comment]
Fantastic. If nothing else that sets up possibilities for conditions/actions/expressions/etc to be put in place for them.

I'm still awaiting Bountysource support to finalize the funds/the timeout to occur so I can move the bounty to this issue properly. I'll update here with the total and new link once the bounty is live.

**Comment by Silver-Streak** [comment]
The bountysource funds have finally been processed. I've opened the bounty on Rysolv (nice to see it fully integrate with Github and update when funds have been applied.)

**Comment by Silver-Streak** [comment]
This has neared another birthday, and people are asking about this again on discord and the forums. 

I've bumped up the bounty on this and the ldtk file support to hopeful renew drive/interest.

**Comment by willianholtz** [comment]
@Silver-Streak 
Sorry for the intrusion. but..
Why is this integration taking so long? Something related to the tool or to GDevelop? Can you answer me? If it's with the tool, why not use Tiled? This is really taking too long, as you said yourself "it's having another birthday", and we don't even have anything experimental!

**Comment by 4ian** [comment]
Basically: we're lacking people to work on this. This needs work, and we don't have enough time.
D8H from the team has been working on improving tilemaps with support for collisions set in Tiled, this will pave the way for native support of the LDTK format.

**Comment by Mix3rn** [comment]
Any new updates ?

**Comment by Silver-Streak** [comment]
No one has approached the project (Or the bounty) for taking on this bounty to bundle the editor. LDTK map support has been fully implemented so you can use it's maps directly without having to go LDTK > Tiled > GDevelop, though.

**Comment by Silver-Streak** [comment]
@daiyam This bounty is over on Rysolv instead of bountysource, so I didn't know if you had seen it or not. With the maps themselves now supported, I wanted to at least draw your attention to it if you had any interest.

**Comment by daiyam** [comment]
I will take a look

**Comment by Silver-Streak** [comment]
Just checking back on this @daiyam, did you get a chance to take a look?

**Comment by daiyam** [comment]
@Silver-Streak Ya a bit, but I didn't had much time since some urgent stuff came. I should be able to re-try soon. 
But I will try a Piskel-like integration.

**Comment by daiyam** [comment]
OK, I was able to spent some time on this. 
First of all, LDtk is heavily integrated with Electron (which is normal).
While, I could make it works in the electron-app, I don't think it possible for the web-app. Is that ok?

**Comment by Silver-Streak** [comment]
Probably a question for @4ian as I know the main dev team are the ones that have had to do all of the bundled app conversions

**Comment by daiyam** [comment]
Another possibility would be to isolate/rewrite the main editor but that would ask lot of works. Not sure if it worths it.

**Comment by Silver-Streak** [comment]
I definitely think we would want to avoid that. I believe The hope would be to have it as close to the original as possible so that version upgrades are as seamless as possible.

**Comment by daiyam** [comment]
I will try to polyfill all the required functions (from nodejs or electron) and see it's doable.

**Comment by FlokiTV** [comment]
still open? i'll do

**Comment by Silver-Streak** [comment]
Hey @FlokiTV

I believe @daiyam is working on this currently. daiyam, if you run into things that make you think you won't be able to proceed, definitely let us know so Floki can take a crack at it.

**Comment by daiyam** [comment]
@FlokiTV Go ahead! You can give it a try.

**Comment by Silver-Streak** [comment]
@FlokiTV if you are wanting to take it on, absolutely take a crack at it.

As a reminder the goal for the integration/bundling is the following:

```
Wrapping LDtk as a react component for GDevelop and adding the react component as a new tab in 
GDevelop 5. If too complex, consider running LDtk as a separate dialog (like Piskel)

Allow for two-way communication between GD5 and LDtk, or at least for edits to existing tilemaps in 
LDtk to update tilemap objects in GD5 (may already be covered by the below PR)
```

There's also a goal to pull in LDTK entity and enum data into variables/objects, but that may need to be a separate PR. If you get through the above and also think that's feasible, let us know.

**Comment by FlokiTV** [comment]
I've been studying the structure of LDTK for a while, honestly I don't see the difficulty of creating a lib to convert LDTK x GDevelop and vice versa. But LDTK has some disconnected things that I don't think would work for GDevelop.

Lately, I've been improving my vanilla javascript knowledge base and I was really interested in this:
```
Building out a custom GDevelop tilemap editor is another option, but that seems to replicate work done elsewhere,
and means that a lot of the development would have to be helmed by the GDevelop contributors long term,
whereas the majority of functional and logical needs behind tilemaps are already covered by LDtk.
```

**Comment by FlokiTV** [comment]
I believe I can't help

**Comment by Silver-Streak** [comment]
Tht is okay, thAnks for looking, @FlokiTV. 

@daiyam, if you are still up for it, it'd be greatly appreciated if you were to take a crack at it.

**Comment by Silver-Streak** [comment]
@daiyam Just checking back in on this to see if you had a chance to look into this any deeper.

**Comment by daiyam** [comment]
No, I haven't. But here the steps, I think, are needed to be done:
- fork LDtk
- add a middle layer between the editor (top) and electron (bottom)
- integrate by implementing the bottom layer for GDevelop

**Comment by Silver-Streak** [comment]
Hello @daiyam , just checking back in on this. Are you still interested in working on this bounty? If so, do you have an ETA?

**Comment by daiyam** [comment]
Ya, when I have some time so no ETA.

**Comment by Silver-Streak** [comment]
As a heads up, I have bumped up this bounty with what I could. It is now at $300 for anyone who is able to implement this as described.

**Comment by DynachDev** [comment]
Hello, sorry to bother but is this being worked on yet?

**Comment by Silver-Streak** [comment]
> Hello, sorry to bother but is this being worked on yet?

All of the parties that had been interested in doing this did not actually move forward with a PR for review or submission.

**Comment by nexiumbiz-debug** [comment]
Hi, checking this before digging in: is the bounty still active, and is the remaining desired scope specifically embedding LDtk as an editor tab/dialog now that .ldtk tilemap support has landed in earlier PRs? If yes, I can start with a scoped implementation plan/PR focused on the integration surface rather than duplicating the existing tilemap import support.

**Comment by Silver-Streak** [comment]
> Hi, checking this before digging in: is the bounty still active, and is the remaining desired scope specifically embedding LDtk as an editor tab/dialog now that .ldtk tilemap support has landed in earlier PRs? If yes, I can start with a scoped implementation plan/PR focused on the integration surface rather than duplicating the existing tilemap import support.

Hello!

Since this was posted, 2 things happend:

* Rysolv died, taking everyone's contributions towards the bounties with it.
* A custom tilemap editor was built and integrated into the platform
While the auto tiling and data features in LDTk are amazing and I'd prefer for it to be available as an option, I don't know if @4ian finds this as a viable path anymore. (4ian please chime in)

 Also there's no longer a viable bounty platform that I'm aware of that could safely transfer funds if people wanted to fund it again.

**Comment by zhaog100** [comment]
I am interested in working on this LDtk integration bounty. This would be a substantial but well-scoped feature addition. I will begin investigating the integration points and will submit an implementation plan.

**Comment by zhaog100** [comment]
Claiming this bounty. I will work on integrating the Level Designer Toolkit as described. Will submit PR shortly.

**Comment by zhaog100** [comment]
I claim this bounty. GitHub: zhaog100 🌶️ (小米辣)

**Comment by Silver-Streak** [comment]
Hey @zhaog100, you are welcome to consider submitting a PR assuming the maintainers approve this, but I wanted to be clear:

All of the funds for this bounty are gone. As the rysolv platform no longer exists and took all bounty funds with it. If you were to work on this item, it would be for $0.
